File: camera.py
import numpy as np


def rotate_up_vector(up_vector, angle, front):
    angle = np.deg2rad(angle)
    front = front / np.linalg.norm(front)
    cos_angle = np.cos(angle)
    sin_angle = np.sin(angle)
    x, y, z = front

    rotation_matrix = np.array([
        [(cos_angle + x * x * (1 - cos_angle)), (x * y * (1 - cos_angle) - z * sin_angle),
         (x * z * (1 - cos_angle) + y * sin_angle)],
        [(y * x * (1 - cos_angle) + z * sin_angle), (cos_angle + y * y * (1 - cos_angle)),
         (y * z * (1 - cos_angle) - x * sin_angle)],
        [(z * x * (1 - cos_angle) - y * sin_angle), (z * y * (1 - cos_angle) + x * sin_angle),
         (cos_angle + z * z * (1 - cos_angle))]
    ])
    up_vector = np.dot(rotation_matrix, up_vector)
    return up_vector


def fix_angles(yaw_degree, pitch_degree, roll_degree):
    yaw_degree = 180.0 - yaw_degree
    if yaw_degree < 0.0:
        yaw_degree = 360.0 + yaw_degree
    elif yaw_degree >= 360.0:
        yaw_degree = yaw_degree - 360.0

    pitch_degree = (-1) * pitch_degree
    roll_degree = (-1) * roll_degree
    return yaw_degree, pitch_degree, roll_degree


class Camera:
    def __init__(self, fov_horizontal, aspect_ratio, near, far):
        self.fov_horizontal = fov_horizontal
        self.aspect_ratio = aspect_ratio
        self.near = near
        self.far = far

        self.position = np.array([0.0, 0.0, 0.0])
        self.up = np.array([0.0, 1.0, 0.0])
        self.target = np.array([0.0, 0.0, 0.0])
        self.direction = self.target - self.position
        self.yaw_degree = 0.0
        self.pitch_degree = 0.0
        self.roll_degree = 0.0

    def set_position(self, position):
        self.position = position

    def set_angles(self, yaw, pitch, roll):
        yaw, pitch, roll = fix_angles(yaw, pitch, roll)
        self.yaw_degree = yaw
        self.pitch_degree = pitch
        self.roll_degree = roll
        self.__update_vectors()

    def get_view_matrix(self):
        eye = self.position
        eye_target = eye - self.target

        z_matrix = eye_target / np.linalg.norm(eye_target)
        x_matrix = np.cross(self.up, z_matrix)
        x_matrix = x_matrix / np.linalg.norm(x_matrix)
        y_matrix = np.cross(z_matrix, x_matrix)
        y_matrix = y_matrix / np.linalg.norm(y_matrix)

        x_translation = np.dot(x_matrix, -eye)
        y_translation = np.dot(y_matrix, -eye)
        z_translation = np.dot(z_matrix, -eye)

        view_matrix = np.array([
            x_matrix[0], y_matrix[0], z_matrix[0], 0,  # col 1
            x_matrix[1], y_matrix[1], z_matrix[1], 0,  # col 2
            x_matrix[2], y_matrix[2], z_matrix[2], 0,  # col 3
            x_translation, y_translation, z_translation, 1  # col 4
        ])

        return view_matrix

    def get_projection_matrix(self):
        f = 1 / np.tan(np.deg2rad(self.fov_horizontal / 2.0))
        projection_matrix = np.array([
            f / self.aspect_ratio, 0.0, 0.0, 0.0,  # col 1
            0.0, f, 0.0, 0.0,  # col 2
            0.0, 0.0, (self.far + self.near) / (self.near - self.far), -1.0,  # col 3
            0.0, 0.0, (2 * self.far * self.near) / (self.near - self.far), 0.0  # col 4
        ])

        return projection_matrix

    def __update_vectors(self):
        self.__update_direction()
        self.__update_up()
        self.__update_target()

    def __update_direction(self):
        yaw_radians = np.deg2rad(self.yaw_degree)
        pitch_radians = np.deg2rad(self.pitch_degree)

        direction = np.zeros(3)
        direction[0] = np.cos(yaw_radians) * np.cos(pitch_radians)
        direction[1] = np.sin(pitch_radians)
        direction[2] = np.sin(yaw_radians) * np.cos(pitch_radians)

        self.direction = direction / np.linalg.norm(direction)

    def __update_up(self):
        self.up = rotate_up_vector(self.up, self.roll_degree, self.direction)

    def __update_target(self):
        self.target = self.position + self.direction

File: app/src/main/java/com/example/peaksrecognition/Camera.java
package com.example.peaksrecognition;

public class Camera {
    private final double fovHorizontal;
    private final float aspectRatio;
    private final float near;
    private final float far;
    private final double[] upVector;
    private final double[] position;
    private final double[] targetVector;
    private double[] angles;
    private double[] directionVector;

    public Camera(double fovHorizontal, float aspectRatio, float near, float far) {
        this.fovHorizontal = fovHorizontal;
        this.aspectRatio = aspectRatio;
        this.near = near;
        this.far = far;

        position = new double[]{0.0, 0.0, 0.0};
        angles = new double[]{0.0, 0.0, 0.0};
        upVector = new double[]{0.0, 1.0, 0.0};
        targetVector = new double[]{0.0, 0.0, 0.0};
        directionVector = new double[]{0.0, 0.0, 0.0};
    }

    public void setPosition(double x, double y, double z) {
        position[0] = x;
        position[1] = y;
        position[2] = z;
    }

    public void setAngles(double yawDegree, double pitchDegree, double rollDegree) {
        angles = fixAngles(yawDegree, pitchDegree, rollDegree);
        updateVectors();
    }

    public float[] getProjectionMatrix() {
        float f = 1.0f / (float) Math.tan(Math.toRadians(fovHorizontal / 2.0));

        return new float[]{
                (f / aspectRatio), 0.0f, 0.0f, 0.0f,  // col 1
                0.0f, f, 0.0f, 0.0f,  // col 2
                0.0f, 0.0f, (far + near) / (near - far), -1.0f,  // col 3
                0.0f, 0.0f, (2.0f * far * near) / (near - far), 0.0f  // col 4
        };
    }

    public float[] getViewMatrix() {
        double[] eyeTarget = new double[]{
                position[0] - targetVector[0],
                position[1] - targetVector[1],
                position[2] - targetVector[2]
        };
        double[] negatedEye = new double[]{
                (-1.0) * position[0],
                (-1.0) * position[1],
                (-1.0) * position[2]
        };
        double[] zMatrix = Utility.normalizeVectorDouble(eyeTarget, 3);
        double[] xMatrix = Utility.normalizeVectorDouble(Utility.crossProduct3ElementsVector(upVector, zMatrix), 3);
        double[] yMatrix = Utility.crossProduct3ElementsVector(zMatrix, xMatrix);

        float xTranslation = (float) Utility.dotProduct(xMatrix, 3, negatedEye, 3);
        float yTranslation = (float) Utility.dotProduct(yMatrix, 3, negatedEye, 3);
        float zTranslation = (float) Utility.dotProduct(zMatrix, 3, negatedEye, 3);

        return new float[]{
                -(float) xMatrix[0], (float) yMatrix[0], (float) zMatrix[0], 0.0f,  // col 1
                -(float) xMatrix[1], (float) yMatrix[1], (float) zMatrix[1], 0.0f,  // col 2
                -(float) xMatrix[2], (float) yMatrix[2], (float) zMatrix[2], 0.0f,  // col 3
                -xTranslation, yTranslation, zTranslation, 1.0f,  // col 1
        };
    }

    private double[] fixAngles(double yawDegree, double pitchDegree, double rollDegree) {
        yawDegree = 180.0 - yawDegree;
        if (yawDegree < 0.0) {
            yawDegree = 360.0 + yawDegree;
        } else if (yawDegree >= 360.0) {
            yawDegree = yawDegree - 360.0;
        }

        pitchDegree = (-1.0) * pitchDegree;
        rollDegree = (-1.0) * rollDegree;

        return new double[]{yawDegree, pitchDegree, rollDegree};
    }

    private void updateVectors() {

        updateDirectionVector();
        updateUpVector();
        updateTargetVector();
    }

    private void updateDirectionVector() {
        double yawRadians = Math.toRadians(angles[0]);
        double pitchRadians = Math.toRadians(angles[1]);

        double[] direction = new double[3];
        direction[0] = Math.cos(yawRadians) * Math.cos(pitchRadians);
        direction[1] = Math.sin(pitchRadians);
        direction[2] = Math.sin(yawRadians) * Math.cos(pitchRadians);

        double magnitude = Math.sqrt(direction[0] * direction[0] + direction[1] * direction[1] + direction[2] * direction[2]);

        direction[0] /= magnitude;
        direction[1] /= magnitude;
        direction[2] /= magnitude;

        directionVector = direction;
    }

    private void updateUpVector() {
        double rollRadians = Math.toRadians(angles[2]);
        double cosAngle = Math.cos(rollRadians);
        double sinAngle = Math.sin(rollRadians);

        double x = directionVector[0];
        double y = directionVector[1];
        double z = directionVector[2];

        double[][] rotationMatrix = new double[][]{
                {(cosAngle + x * x * (1 - cosAngle)), (x * y * (1 - cosAngle) - z * sinAngle), (x * z * (1 - cosAngle) + y * sinAngle)},
                {(y * x * (1 - cosAngle) + z * sinAngle), (cosAngle + y * y * (1 - cosAngle)), (y * z * (1 - cosAngle) - x * sinAngle)},
                {(z * x * (1 - cosAngle) - y * sinAngle), (z * y * (1 - cosAngle) + x * sinAngle), (cosAngle + z * z * (1 - cosAngle))}
        };

        for (int i = 0; i < 3; ++i) {
            upVector[i] = rotationMatrix[i][0] * upVector[0] + rotationMatrix[i][1] * upVector[1] + rotationMatrix[i][2] * upVector[2];
        }
    }

    private void updateTargetVector() {
        for (int i = 0; i < 3; ++i) {
            targetVector[i] = position[i] + directionVector[i];
        }
    }
}
